servers/v3-proxy-api/src/routes/validations/SendActionValidators.ts

```diff
@@ -0,0 +1,662 @@
+ * Should be used with a "Sanitize" class to return sanitized
+ * data which conforms to the expected action schema. See `HasItemIdOrUrl`
+ * for example usage.
+ */
+function MaybeHasTags<TBase extends ActionSanitizable>(Base: TBase) {
+  return class MaybeHasTags extends Base {
+    readonly tags: string[] | undefined;
+    constructor(...args: any[]) {
+      super(...args);
+      if ('tags' in this.input) {
+        const tags = this.input.tags.split(',');
+        validTagsOrError(tags);
+        this.tags = tags;
+      }
+    }
+  };
+}
+
+/**
+ * Class mixin for input validation. Use to construct a Validation/Sanitizer
+ * class which has the rule: there must be a valid `tags` field
+ * in the input passed to the constructor. Checks to ensure
+ * that the tags field is a comma-separated string with at least one element,
+ * and none of the elements are empty. Transforms the input to an array of tags
+ * if the validation succeeds.
+ *
+ * Must be called with a Class that implements or extends the
+ * ActionSanitizable type (constructor with a single argument, `input`,
+ * which contains an object minimally with a valid key-value pair for 'action').
+ *
+ * Can be chained with other mixins to add additional validation rules.
+ * Should be used with a "Sanitize" class to return sanitized
+ * data which conforms to the expected action schema. See `HasItemIdOrUrl`
+ * for example usage.
+ */
+function HasValidTags<TBase extends ActionSanitizable>(Base: TBase) {
+  return class ValidTags extends Base {
+    tags: string[];
+    constructor(...args: any[]) {
+      super(...args);
+      if (!('tags' in this.input)) {
+        const error: ArrayFieldError = {
+          type: 'array_field',
+          path: 'tags',
+          msg: `Action must have tags field`,
+          value: this.input.tags,
+        };
+        throw new InputValidationError(error);
+      }
+      const tags = this.input.tags.split(',');
+      validTagsOrError(tags);
+      this.tags = tags;
+    }
+  };
+}
+
+/**
+ * Class mixin for input validation. Use to construct a Validation/Sanitizer
+ * class which has the rule: there must be a valid `new_tag` field
+ * in the input passed to the constructor. It must be a non-empty string.
+ *
+ * Must be called with a Class that implements or extends the
+ * ActionSanitizable type (constructor with a single argument, `input`,
+ * which contains an object minimally with a valid key-value pair for 'action').
+ *
+ * Can be chained with other mixins to add additional validation rules.
+ * Should be used with a "Sanitize" class to return sanitized
+ * data which conforms to the expected action schema. See `HasItemIdOrUrl`
+ * for example usage.
+ */
+function HasNewTag<TBase extends ActionSanitizable>(Base: TBase) {
+  return class extends Base {
+    readonly newTag: string;
+    constructor(...args: any[]) {
+      super(...args);
+      nonEmptyStringPropOrError(this.input, 'new_tag');
```

Comment:
🤣 js. oh js. if only there was a simple not null, not undefined, not empty built in :shakes--fist: